testTop/pyTestsApp/gateway_tests/conftest.py

```diff
@@ -0,0 +1,993 @@
+        context_set_env("EPICS_CA_ADDR_LIST", address_list),
+    ):
+        reset_libca()
+        yield
+
+
+@contextlib.contextmanager
+def context_set_env(key: str, value: Any):
+    """Context manager to set - and then reset - an environment variable."""
+    orig_value = os.environ.get(key, None)
+    try:
+        os.environ[key] = str(value)
+        yield orig_value
+    finally:
+        if orig_value is not None:
+            os.environ[key] = orig_value
+
+
+def is_pyepics_libca_initialized() -> bool:
+    """Has pyepics' LIBCA been initialized?"""
+    return epics.ca.libca not in (None, epics.ca._LIBCA_FINALIZED)
+
+
+def connect_libca() -> None:
+    if not is_pyepics_libca_initialized():
+        gc.collect()
+        epics.ca.initialize_libca()
+
+
+def disconnect_libca() -> None:
+    if is_pyepics_libca_initialized():
+        epics.ca.finalize_libca()
+        epics.ca.clear_cache()
+
+
+def reset_libca() -> None:
+    disconnect_libca()
+    connect_libca()
+
+
+@contextlib.contextmanager
+def gateway_channel_access_env(port: int = config.default_gateway_port):
+    """Set the environment up for communication solely with the spawned gateway."""
+    with (
+        context_set_env("EPICS_CA_AUTO_ADDR_LIST", "NO"),
+        context_set_env("EPICS_CA_ADDR_LIST", f"localhost:{port}"),
+    ):
+        reset_libca()
+        yield
+
+
+@contextlib.contextmanager
+def ioc_channel_access_env(port: int = config.default_ioc_port):
+    """Set the environment up for communication solely with the spawned IOC."""
+    with (
+        context_set_env("EPICS_CA_AUTO_ADDR_LIST", "NO"),
+        context_set_env("EPICS_CA_ADDR_LIST", f"localhost:{port}"),
+    ):
+        reset_libca()
+        yield
+
+
+@dataclasses.dataclass
+class EnvironmentInfo:
+    """
+    Test environment information.
+
+    An instance of this is used as a fixture for ``standard_env``.
+    """
+
+    access: str
+    pvlist: str
+    db_file: str
+    dbd_file: Optional[str]
+
+
+@contextlib.contextmanager
+def file_test_environment(
+    access: str,
+    pvlist: str,
+    db_file: str,
+    dbd_file: Optional[str] = None,
+) -> Generator[EnvironmentInfo, None, None]:
+    """
+    Test environment using already-existing files on disk.
+
+    Parameters
+    ----------
+    access : str
+        The access rights filename.
+    pvlist : str
+        The pvlist filename.
+    db_file : str
+        Path to the IOC database.
+    dbd_file : str, optional
+        Path to the IOC database definition.  Defaults to using the database
+        definition provided with epics-base/softIoc.
+    """
+    with (
+        run_gateway(access=access, pvlist=pvlist),
+        run_ioc(db_file=db_file, dbd_file=dbd_file),
+        local_channel_access(),
+    ):
+        yield EnvironmentInfo(
+            access=access,
+            pvlist=pvlist,
+            db_file=db_file,
+            dbd_file=dbd_file,
+        )
+
+
+@pytest.fixture(scope="function")
```

Comment:
I note usage of pytest [fixtures](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html) as I found these super confusing when I first saw eg.

```py
def test_cs_studio_value_and_prop_monitor(standard_env: conftest.EnvironmentInfo):
```

without understanding where this magical `standard_env` argument comes from.